File: reveal-sc2-opponent-main/core/integrity_sweep.py
"""
core.integrity_sweep -- Stage 5 of STAGE_DATA_INTEGRITY_ROADMAP.

Walks ``data/`` at every backend / watcher / scanner start, surfaces
orphaned ``.tmp_*.json`` files, detects corrupt live files, and places
recovery candidates in ``data/.recovery/<basename>/`` for the user
to review and apply through the Diagnostics page (or the CLI).

Design notes
------------

* **Read-only.** The sweep never publishes a recovery automatically.
  All it does is move/copy files into the ``data/.recovery/<base>/``
  staging directory. A user clicks "Apply recovery" in the SPA (or
  runs ``python -m core.integrity_sweep --apply <candidate>``) to
  promote the candidate.
* **Idempotent.** Running the sweep N times against the same state
  produces the same staging output -- the candidate file is named
  ``<basename>-<UTC>.json`` so we never overwrite a previous run's
  candidate. Safe to invoke on every process start.
* **PII-safe.** Only counts and basenames are logged at INFO. Pulse
  IDs / opponent names never appear in log lines.
* **Cross-language friendly.** Mirrors the JS-side helpers in
  ``stream-overlay-backend/lib/integrity_sweep.js`` byte-for-byte on
  the staging-directory layout so either implementation can serve
  recovery candidates to the SPA.

Triage tiers per live file
--------------------------

1. Live parses cleanly + matches floor    -> OK
2. Live parses cleanly + below floor      -> CORRUPT_SMALL
3. Live fails to parse                    -> CORRUPT_UNPARSEABLE
4. Live missing                            -> MISSING

For tiers 2/3/4 the sweep looks for a candidate in this order:

* a ``.tmp_*.json`` orphan in the same dir that parses cleanly and
  has a key count >= the live file's
* the matching ``<basename>.bak`` snapshot
* the newest file in ``data/.recovery/<basename>/`` (a previous-
  run's candidate that was never applied)

Exactly the first hit is staged. If none of them are usable the
sweep raises a WARN and leaves the live file alone.

Exposed API
-----------

* :func:`run_sweep(data_dir, *, now=...)` -> :class:`SweepReport`
* :func:`apply_candidate(candidate_path, target_path)` -> ``None``
* CLI ``python -m core.integrity_sweep --data-dir DIR``
* CLI ``python -m core.integrity_sweep --apply CANDIDATE``
"""

from __future__ import annotations

import argparse
import dataclasses
import datetime as _dt
import json
import logging
import os
import shutil
import sys
import time
from typing import Any, Dict, List, Optional

# ---------------------------------------------------------------------------
# Tunables
# ---------------------------------------------------------------------------
DEFAULT_TMP_AGE_THRESHOLD_SEC = 300  # 5 min -- young .tmp may be a live write
RECOVERY_DIR_NAME = ".recovery"
ORPHAN_QUARANTINE_PREFIX = "recovery-orphans-"
TRACKED_BASENAMES = (
    "MyOpponentHistory.json",
    "meta_database.json",
    "custom_builds.json",
    "profile.json",
    "config.json",
)


logger = logging.getLogger("integrity_sweep")


# ---------------------------------------------------------------------------
# UTC helper -- timezone-aware, never deprecated
# ---------------------------------------------------------------------------
def _utc_stamp() -> str:
    return _dt.datetime.now(_dt.timezone.utc).strftime("%Y%m%dT%H%M%SZ")


# ---------------------------------------------------------------------------
# Floors mirror core.atomic_io.FILE_FLOORS but kept in a local copy so
# this module imports cleanly even on a partial install.
# ---------------------------------------------------------------------------
def _resolve_floor(basename: str) -> int:
    try:
        from core.atomic_io import FILE_FLOORS  # late import: optional dep
        return int(FILE_FLOORS.get(basename, 0) or 0)
    except Exception:  # noqa: BLE001
        return 0


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------
def _read_json_bytes(path: str) -> Optional[Any]:
    try:
        with open(path, "rb") as f:
            raw = f.read()
        if raw.startswith(b"\xef\xbb\xbf"):
            raw = raw[3:]
        raw = raw.strip(b" \t\r\n\x00")
        if not raw:
            return None
        return json.loads(raw.decode("utf-8"))
    except Exception:  # noqa: BLE001
        return None


def _key_count(parsed: Any) -> int:
    return len(parsed) if isinstance(parsed, dict) else 0


def _atomic_publish_bytes(path: str, payload: bytes) -> None:
    """Bytes-level atomic publish used by apply_candidate.

    We deliberately do not delegate to ``core.atomic_io.atomic_write_json``
    here: the candidate may already be a complete on-disk file we just
    want to swap in. We do the same fsync+rename dance with a fresh
    temp so the swap survives a kill mid-publish.
    """
    parent = os.path.dirname(path) or "."
    os.makedirs(parent, exist_ok=True)
    tmp_path = os.path.join(parent, ".tmp_apply_recovery.json")
    try:
        with open(tmp_path, "wb") as f:
            f.write(payload)
            f.flush()
            os.fsync(f.fileno())
        # .bak the old live file before publishing so safe-read fallbacks
        # still work if the candidate later turns out to be wrong.
        if os.path.exists(path):
            try:
                shutil.copy2(path, path + ".bak")
            except OSError:
                pass
        os.replace(tmp_path, path)
    except Exception:
        try:
            os.unlink(tmp_path)
        except OSError:
            pass
        raise


# ---------------------------------------------------------------------------
# Findings dataclass
# ---------------------------------------------------------------------------
@dataclasses.dataclass
class FileFinding:
    basename: str
    live_path: str
    status: str                        # "ok" | "missing" | "corrupt_small" | "corrupt_unparseable"
    live_keys: int = 0
    candidate_path: Optional[str] = None  # path under data/.recovery/<base>/
    candidate_keys: int = 0
    candidate_source: Optional[str] = None  # "orphan" | "bak" | "stale_recovery" | None
    notes: List[str] = dataclasses.field(default_factory=list)

    def to_dict(self) -> Dict[str, Any]:
        return dataclasses.asdict(self)


@dataclasses.dataclass
class SweepReport:
    data_dir: str
    timestamp: str
    findings: List[FileFinding]
    orphans_seen: List[str]
    orphans_aged: List[str]            # those past the age threshold
    candidates_staged: List[str]
    warnings: List[str]

    def to_dict(self) -> Dict[str, Any]:
        return {
            "data_dir": self.data_dir,
            "timestamp": self.timestamp,
            "findings": [f.to_dict() for f in self.findings],
            "orphans_seen": list(self.orphans_seen),
            "orphans_aged": list(self.orphans_aged),
            "candidates_staged": list(self.candidates_staged),
            "warnings": list(self.warnings),
        }


# ---------------------------------------------------------------------------
# Discovery
# ---------------------------------------------------------------------------
def _discover_orphans(data_dir: str, now: float, age_threshold_sec: float) -> Dict[str, Any]:
    """Return ``{seen: [...], aged: [...]}``."""
    seen: List[str] = []
    aged: List[str] = []
    if not os.path.isdir(data_dir):
        return {"seen": seen, "aged": aged}
    for name in os.listdir(data_dir):
        if not (name.startswith(".tmp_") and name.endswith(".json")):
            continue
        path = os.path.join(data_dir, name)
        if not os.path.isfile(path):
            continue
        seen.append(path)
        try:
            mtime = os.path.getmtime(path)
        except OSError:
            mtime = 0.0
        if now - mtime >= age_threshold_sec:
            aged.append(path)
    return {"seen": seen, "aged": aged}


def _newest_recovery_candidate(data_dir: str, basename: str) -> Optional[str]:
    """Newest stale candidate left over from a previous sweep, if any."""
    rdir = os.path.join(data_dir, RECOVERY_DIR_NAME, basename)
    if not os.path.isdir(rdir):
        return None
    files = []
    for name in os.listdir(rdir):
        if not name.endswith(".json"):
            continue
        path = os.path.join(rdir, name)
        if os.path.isfile(path):
            files.append(path)
    if not files:
        return None
    files.sort(key=lambda p: os.path.getmtime(p), reverse=True)
    return files[0]


def _stage_candidate(
    data_dir: str,
    basename: str,
    source_path: str,
    *,
    source_label: str,
) -> str:
    """Copy ``source_path`` into ``data/.recovery/<basename>/<base>-<UTC>.json``.

    Returns the absolute candidate path. Existing candidate files are
    not overwritten -- we always create a new timestamped name.
    """
    rdir = os.path.join(data_dir, RECOVERY_DIR_NAME, basename)
    os.makedirs(rdir, exist_ok=True)
    stem = os.path.splitext(basename)[0]
    candidate_path = os.path.join(rdir, f"{stem}-{_utc_stamp()}-{source_label}.json")
    shutil.copy2(source_path, candidate_path)
    return candidate_path


# ---------------------------------------------------------------------------
# Per-file triage
# ---------------------------------------------------------------------------
def _triage_one(
    data_dir: str,
    basename: str,
    orphan_pool: List[str],
) -> FileFinding:
    """Inspect one tracked file and decide whether a candidate is needed."""
    live_path = os.path.join(data_dir, basename)
    finding = FileFinding(basename=basename, live_path=live_path, status="ok")
    floor = _resolve_floor(basename)

    if not os.path.exists(live_path):
        finding.status = "missing"
    else:
        parsed = _read_json_bytes(live_path)
        if parsed is None:
            finding.status = "corrupt_unparseable"
        else:
            finding.live_keys = _key_count(parsed)
            if floor and finding.live_keys < floor:
                finding.status = "corrupt_small"

    if finding.status == "ok":
        return finding

    # Need a candidate. Look in the orphan pool first.
    best_orphan: Optional[str] = None
    best_orphan_keys = -1
    for orphan in orphan_pool:
        parsed = _read_json_bytes(orphan)
        if not isinstance(parsed, dict):
            continue
        kc = len(parsed)
        if kc <= finding.live_keys:
            continue
        if kc > best_orphan_keys:
            best_orphan_keys = kc
            best_orphan = orphan
    if best_orphan is not None:
        candidate = _stage_candidate(
            data_dir, basename, best_orphan, source_label="orphan"
        )
        finding.candidate_path = candidate
        finding.candidate_keys = best_orphan_keys
        finding.candidate_source = "orphan"
        return finding

    # Try .bak.
    bak_path = live_path + ".bak"
    bak_parsed = _read_json_bytes(bak_path)
    if isinstance(bak_parsed, dict) and len(bak_parsed) > finding.live_keys:
        candidate = _stage_candidate(
            data_dir, basename, bak_path, source_label="bak"
        )
        finding.candidate_path = candidate
        finding.candidate_keys = len(bak_parsed)
        finding.candidate_source = "bak"
        return finding

    # Stale recovery from a previous sweep.
    stale = _newest_recovery_candidate(data_dir, basename)
    if stale and stale != finding.candidate_path:
        stale_parsed = _read_json_bytes(stale)
        if isinstance(stale_parsed, dict) and len(stale_parsed) > finding.live_keys:
            finding.candidate_path = stale
            finding.candidate_keys = len(stale_parsed)
            finding.candidate_source = "stale_recovery"
            return finding

    finding.notes.append("no usable candidate found")
    return finding


# ---------------------------------------------------------------------------
# Public sweep entry point
# ---------------------------------------------------------------------------
def run_sweep(
    data_dir: str,
    *,
    now: Optional[float] = None,
    tmp_age_threshold_sec: float = DEFAULT_TMP_AGE_THRESHOLD_SEC,
) -> SweepReport:
    """Walk ``data_dir`` once and stage any recovery candidates we find."""
    data_dir = os.path.abspath(data_dir)
    if now is None:
        now = time.time()
    report = SweepReport(
        data_dir=data_dir,
        timestamp=_utc_stamp(),
        findings=[],
        orphans_seen=[],
        orphans_aged=[],
        candidates_staged=[],
        warnings=[],
    )
    if not os.path.isdir(data_dir):
        report.warnings.append(f"data_dir does not exist: {data_dir}")
        return report

    orphans = _discover_orphans(data_dir, now=now, age_threshold_sec=tmp_age_threshold_sec)
    report.orphans_seen = sorted(orphans["seen"])
    report.orphans_aged = sorted(orphans["aged"])

    # Only consider AGED orphans for recovery -- a young one might be a
    # live write in progress by another process.
    pool = list(orphans["aged"])

    for basename in TRACKED_BASENAMES:
        finding = _triage_one(data_dir, basename, pool)
        report.findings.append(finding)
        if finding.candidate_path:
            report.candidates_staged.append(finding.candidate_path)
        if finding.status != "ok" and not finding.candidate_path:
            report.warnings.append(
                f"{basename}: {finding.status} -- no usable candidate"
            )

    if report.candidates_staged:
        logger.info("[integrity] candidates: %d", len(report.candidates_staged))
    elif any(f.status != "ok" for f in report.findings):
        logger.warning(
            "[integrity] %d file(s) in non-ok state but no candidates staged",
            sum(1 for f in report.findings if f.status != "ok"),
        )
    else:
        logger.info("[integrity] OK")

    return report


# ---------------------------------------------------------------------------
# Apply
# ---------------------------------------------------------------------------
def apply_candidate(candidate_path: str, target_path: str) -> None:
    """Promote ``candidate_path`` to ``target_path`` via atomic publish.

    Validates with Stage 4 gate logic (a corrupt candidate raises
    DataIntegrityError; live file is unchanged).

    Args:
        candidate_path: Path under ``data/.recovery/<base>/``.
        target_path:   Live data/<base> path to overwrite.
    """
    if not os.path.exists(candidate_path):
        raise FileNotFoundError(f"candidate not found: {candidate_path}")
    parsed = _read_json_bytes(candidate_path)
    if not isinstance(parsed, dict):
        from core.atomic_io import DataIntegrityError
        raise DataIntegrityError(
            f"apply_candidate: {candidate_path} did not parse to a dict"
        )
    floor = _resolve_floor(os.path.basename(target_path))
    if floor and len(parsed) < floor:
        from core.atomic_io import DataIntegrityError
        raise DataIntegrityError(
            f"apply_candidate: candidate has {len(parsed)} keys "
            f"(floor={floor}); refusing to apply"
        )

    with open(candidate_path, "rb") as f:
        payload = f.read()
    _atomic_publish_bytes(target_path, payload)


# ---------------------------------------------------------------------------
# CLI
# ---------------------------------------------------------------------------
def _build_argparser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
        description=(
            "Stage 5 -- scan data/ for orphaned tmp files and corrupt "
            "live files; stage recovery candidates."
        )
    )
    parser.add_argument("--data-dir", default=None,
                        help="Directory to scan (default: <repo>/data).")
    parser.add_argument("--apply", default=None,
                        help="Promote the given candidate path to its live target.")
    parser.add_argument("--target", default=None,
                        help="Explicit live target (only with --apply).")
    parser.add_argument("--json", action="store_true",
                        help="Emit the report as JSON on stdout.")
    parser.add_argument("-v", "--verbose", action="store_true")
    return parser


def _default_data_dir() -> str:
    return os.path.abspath(os.path.join(
        os.path.dirname(os.path.abspath(__file__)), os.pardir, "data"
    ))


def main(argv: Optional[List[str]] = None) -> int:
    args = _build_argparser().parse_args(argv)
    logging.basicConfig(
        level=logging.DEBUG if args.verbose else logging.INFO,
        format="%(asctime)s %(levelname)s %(name)s: %(message)s",
    )
    if args.apply:
        candidate = os.path.abspath(args.apply)
        if args.target:
            target = os.path.abspath(args.target)
        else:
            # Default target: the live file with the same basename in data/.
            data_dir = os.path.abspath(args.data_dir or _default_data_dir())
            base = os.path.basename(candidate)
            # Strip trailing -<UTC>-<source>.json -> "<base>.json"
            stem, _ext = os.path.splitext(base)
            head = stem.split("-")[0]
            target = os.path.join(data_dir, head + ".json")
        try:
            apply_candidate(candidate, target)
        except Exception as exc:  # noqa: BLE001
            logger.error("apply_candidate failed: %s", exc)
            return 2
        print(f"applied {candidate} -> {target}")
        return 0

    data_dir = os.path.abspath(args.data_dir or _default_data_dir())
    report = run_sweep(data_dir)
    if args.json:
        print(json.dumps(report.to_dict(), indent=2, default=str))
    else:
        for f in report.findings:
            line = f"{f.basename:24s}  {f.status:22s}  keys={f.live_keys}"
            if f.candidate_path:
                line += f"  -> candidate ({f.candidate_source}, {f.candidate_keys} keys)"
            print(line)
        for w in report.warnings:
            print("WARN:", w)
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: reveal-sc2-opponent-main/stream-overlay-backend/lib/integrity_sweep.js
// @ts-check
/**
 * lib/integrity_sweep.js -- Stage 5 of STAGE_DATA_INTEGRITY_ROADMAP.
 *
 * Mirrors core/integrity_sweep.py byte-for-byte on the staging-
 * directory layout (data/.recovery/<basename>/<base>-<UTC>-<src>.json)
 * so either implementation can drive the SPA's "Apply recovery" UI.
 *
 * The sweep:
 *   1. Walks data/ for `.tmp_*.json` orphans older than 5 minutes.
 *   2. For each tracked basename, decides ok / corrupt-small /
 *      corrupt-unparseable / missing.
 *   3. For non-ok cases, stages a candidate file in
 *      data/.recovery/<basename>/. Never auto-publishes.
 *
 * The matching POST /api/recovery/apply endpoint takes a candidate
 * path, validates via Stage 4 gate, and atomic-replaces the live
 * file. Lives in routes/recovery.js so the diagnostics router stays
 * focused on read-only checks.
 *
 * Engineering preamble compliance:
 *   - JSDoc + tsc --checkJs strict.
 *   - Function size <= 30 lines.
 *   - No magic constants -- all knobs are named module-level constants.
 *   - PII-safe logging: only counts and basenames at INFO.
 */

'use strict';

const fs = require('fs');
const path = require('path');

const TRACKED_BASENAMES = Object.freeze([
  'MyOpponentHistory.json',
  'meta_database.json',
  'custom_builds.json',
  'profile.json',
  'config.json',
]);
const RECOVERY_DIR_NAME = '.recovery';
const DEFAULT_TMP_AGE_THRESHOLD_SEC = 300;
// Mirrors core.atomic_io.FILE_FLOORS -- kept in sync via Stage 6 schema
// versioning. A drift will fail the cross-language consistency test.
const FILE_FLOORS = Object.freeze({
  'MyOpponentHistory.json': 100,
  'meta_database.json':     50,
  'custom_builds.json':     0,
  'profile.json':           1,
  'config.json':            5,
});

function _utcStamp() {
  return new Date().toISOString().replace(/[-:.]/g, '').slice(0, 15) + 'Z';
}

function _readJsonOrNull(filePath) {
  try {
    if (!fs.existsSync(filePath)) return null;
    let raw = fs.readFileSync(filePath, 'utf8');
    if (raw.charCodeAt(0) === 0xFEFF) raw = raw.slice(1);
    raw = raw.replace(/[\s\x00]+$/, '');
    if (!raw) return null;
    return JSON.parse(raw);
  } catch (_err) {
    return null;
  }
}

function _keyCount(parsed) {
  return parsed && typeof parsed === 'object' && !Array.isArray(parsed)
    ? Object.keys(parsed).length
    : 0;
}

/**
 * Walk data/ for `.tmp_*.json` orphans. Aged orphans are ones at
 * least `ageThresholdSec` seconds old -- younger ones may be a live
 * write in progress by another process and we don't touch them.
 *
 * @param {string} dataDir
 * @param {number} now Unix epoch seconds.
 * @param {number} ageThresholdSec
 * @returns {{ seen: string[], aged: string[] }}
 */
function _discoverOrphans(dataDir, now, ageThresholdSec) {
  /** @type {string[]} */ const seen = [];
  /** @type {string[]} */ const aged = [];
  if (!fs.existsSync(dataDir) || !fs.statSync(dataDir).isDirectory()) {
    return { seen, aged };
  }
  for (const name of fs.readdirSync(dataDir)) {
    if (!name.startsWith('.tmp_') || !name.endsWith('.json')) continue;
    const p = path.join(dataDir, name);
    if (!fs.statSync(p).isFile()) continue;
    seen.push(p);
    const ageSec = now - (fs.statSync(p).mtimeMs / 1000);
    if (ageSec >= ageThresholdSec) aged.push(p);
  }
  return { seen, aged };
}

/**
 * Newest stale candidate (from a previous sweep) for `basename`,
 * or null if none. Used as a third-tier fallback.
 *
 * @param {string} dataDir
 * @param {string} basename
 * @returns {string|null}
 */
function _newestStaleCandidate(dataDir, basename) {
  const rdir = path.join(dataDir, RECOVERY_DIR_NAME, basename);
  if (!fs.existsSync(rdir) || !fs.statSync(rdir).isDirectory()) return null;
  const files = fs.readdirSync(rdir)
    .filter((n) => n.endsWith('.json'))
    .map((n) => path.join(rdir, n))
    .filter((p) => fs.statSync(p).isFile())
    .sort((a, b) => fs.statSync(b).mtimeMs - fs.statSync(a).mtimeMs);
  return files.length ? files[0] : null;
}

/**
 * Copy `sourcePath` into data/.recovery/<basename>/<stem>-<UTC>-<src>.json.
 * Never overwrites; always emits a fresh timestamped file.
 *
 * @param {string} dataDir
 * @param {string} basename
 * @param {string} sourcePath
 * @param {string} sourceLabel "orphan" | "bak" | "stale_recovery"
 * @returns {string} absolute candidate path
 */
function _stageCandidate(dataDir, basename, sourcePath, sourceLabel) {
  const rdir = path.join(dataDir, RECOVERY_DIR_NAME, basename);
  fs.mkdirSync(rdir, { recursive: true });
  const stem = basename.replace(/\.json$/, '');
  const candidate = path.join(rdir, `${stem}-${_utcStamp()}-${sourceLabel}.json`);
  fs.copyFileSync(sourcePath, candidate);
  return candidate;
}

/**
 * Inspect one tracked file and stage a candidate if needed.
 *
 * @param {string} dataDir
 * @param {string} basename
 * @param {string[]} orphanPool
 * @returns {{
 *   basename: string,
 *   live_path: string,
 *   status: string,
 *   live_keys: number,
 *   candidate_path: string | null,
 *   candidate_keys: number,
 *   candidate_source: string | null,
 *   notes: string[]
 * }}
 */
function _triageOne(dataDir, basename, orphanPool) {
  const livePath = path.join(dataDir, basename);
  const finding = {
    basename,
    live_path: livePath,
    status: 'ok',
    live_keys: 0,
    candidate_path: /** @type {string|null} */ (null),
    candidate_keys: 0,
    candidate_source: /** @type {string|null} */ (null),
    notes: /** @type {string[]} */ ([]),
  };
  const floor = FILE_FLOORS[basename] || 0;

  if (!fs.existsSync(livePath)) {
    finding.status = 'missing';
  } else {
    const parsed = _readJsonOrNull(livePath);
    if (parsed === null) {
      finding.status = 'corrupt_unparseable';
    } else {
      finding.live_keys = _keyCount(parsed);
      if (floor && finding.live_keys < floor) finding.status = 'corrupt_small';
    }
  }
  if (finding.status === 'ok') return finding;

  let bestOrphan = null;
  let bestKeys = -1;
  for (const orphan of orphanPool) {
    const parsed = _readJsonOrNull(orphan);
    if (!parsed || typeof parsed !== 'object' || Array.isArray(parsed)) continue;
    const kc = Object.keys(parsed).length;
    if (kc <= finding.live_keys) continue;
    if (kc > bestKeys) { bestKeys = kc; bestOrphan = orphan; }
  }
  if (bestOrphan) {
    finding.candidate_path = _stageCandidate(dataDir, basename, bestOrphan, 'orphan');
    finding.candidate_keys = bestKeys;
    finding.candidate_source = 'orphan';
    return finding;
  }

  const bakPath = livePath + '.bak';
  const bakParsed = _readJsonOrNull(bakPath);
  if (bakParsed && typeof bakParsed === 'object' && !Array.isArray(bakParsed)
      && Object.keys(bakParsed).length > finding.live_keys) {
    finding.candidate_path = _stageCandidate(dataDir, basename, bakPath, 'bak');
    finding.candidate_keys = Object.keys(bakParsed).length;
    finding.candidate_source = 'bak';
    return finding;
  }

  const stale = _newestStaleCandidate(dataDir, basename);
  if (stale) {
    const sp = _readJsonOrNull(stale);
    if (sp && typeof sp === 'object' && !Array.isArray(sp)
        && Object.keys(sp).length > finding.live_keys) {
      finding.candidate_path = stale;
      finding.candidate_keys = Object.keys(sp).length;
      finding.candidate_source = 'stale_recovery';
      return finding;
    }
  }
  finding.notes.push('no usable candidate found');
  return finding;
}

/**
 * Walk dataDir once and stage any recovery candidates.
 * Pure function -- no logger, callers wrap it for telemetry.
 *
 * @param {string} dataDir
 * @param {{ now?: number, tmpAgeThresholdSec?: number }} [opts]
 * @returns {{
 *   data_dir: string,
 *   timestamp: string,
 *   findings: object[],
 *   orphans_seen: string[],
 *   orphans_aged: string[],
 *   candidates_staged: string[],
 *   warnings: string[]
 * }}
 */
function runSweep(dataDir, opts) {
  const o = opts || {};
  const now = typeof o.now === 'number' ? o.now : Date.now() / 1000;
  const ageThresh = typeof o.tmpAgeThresholdSec === 'number'
    ? o.tmpAgeThresholdSec : DEFAULT_TMP_AGE_THRESHOLD_SEC;
  const abs = path.resolve(dataDir);
  const out = {
    data_dir: abs,
    timestamp: _utcStamp(),
    findings: /** @type {object[]} */ ([]),
    orphans_seen: /** @type {string[]} */ ([]),
    orphans_aged: /** @type {string[]} */ ([]),
    candidates_staged: /** @type {string[]} */ ([]),
    warnings: /** @type {string[]} */ ([]),
  };
  if (!fs.existsSync(abs) || !fs.statSync(abs).isDirectory()) {
    out.warnings.push(`data_dir does not exist: ${abs}`);
    return out;
  }
  const orphans = _discoverOrphans(abs, now, ageThresh);
  out.orphans_seen = [...orphans.seen].sort();
  out.orphans_aged = [...orphans.aged].sort();
  for (const basename of TRACKED_BASENAMES) {
    const f = _triageOne(abs, basename, orphans.aged);
    out.findings.push(f);
    if (f.candidate_path) out.candidates_staged.push(f.candidate_path);
    if (f.status !== 'ok' && !f.candidate_path) {
      out.warnings.push(`${basename}: ${f.status} -- no usable candidate`);
    }
  }
  return out;
}

/**
 * Apply a previously-staged candidate to its live target. Validates
 * via the Stage 4 gate (DataIntegrityError on shrinkage / parse fail);
 * .bak's the live file before swap; uses fsync + atomic rename so a
 * kill mid-publish leaves either the original or the candidate.
 *
 * @param {string} candidatePath
 * @param {string} targetPath
 */
function applyCandidate(candidatePath, targetPath) {
  if (!fs.existsSync(candidatePath)) {
    throw new Error('applyCandidate: candidate not found: ' + candidatePath);
  }
  const parsed = _readJsonOrNull(candidatePath);
  if (!parsed || typeof parsed !== 'object' || Array.isArray(parsed)) {
    const { DataIntegrityError } = require('./atomic-fs');
    throw new DataIntegrityError(
      `applyCandidate: ${candidatePath} did not parse to an object`);
  }
  const floor = FILE_FLOORS[path.basename(targetPath)] || 0;
  if (floor && Object.keys(parsed).length < floor) {
    const { DataIntegrityError } = require('./atomic-fs');
    throw new DataIntegrityError(
      `applyCandidate: candidate has ${Object.keys(parsed).length} keys `
      + `(floor=${floor}); refusing to apply`);
  }
  const payload = fs.readFileSync(candidatePath);
  const parent = path.dirname(targetPath);
  fs.mkdirSync(parent, { recursive: true });
  const tmp = path.join(parent, '.tmp_apply_recovery.json');
  const fd = fs.openSync(tmp, 'w', 0o644);
  try {
    fs.writeSync(fd, payload, 0, payload.length, 0);
    fs.fsyncSync(fd);
  } finally {
    fs.closeSync(fd);
  }
  if (fs.existsSync(targetPath)) {
    try { fs.copyFileSync(targetPath, targetPath + '.bak'); }
    catch (_e) { /* best-effort */ }
  }
  try {
    fs.renameSync(tmp, targetPath);
  } catch (err) {
    try { fs.unlinkSync(tmp); } catch (_e) { /* */ }
    throw err;
  }
}

module.exports = {
  runSweep,
  applyCandidate,
  TRACKED_BASENAMES,
  RECOVERY_DIR_NAME,
  DEFAULT_TMP_AGE_THRESHOLD_SEC,
  FILE_FLOORS,
  // Internals exposed for tests:
  _internals: {
    _discoverOrphans,
    _newestStaleCandidate,
    _stageCandidate,
    _triageOne,
    _readJsonOrNull,
  },
};
